Add Python 3.12 dev container

heathhenley opened the pull request:
Have to admit I haven't used dev containers much, but I've wanted to play with them. 

I added a normal preconfigured 3.12.8 container using one of the templates built into vscode's 'create dev container' commands (hence all the extra comments in the file). I then changed the `postCreateCommand` to install all the dependencies and set some git settings (`post-create.sh`).

Curious if someone else can try to spin it up and how it works for them. I was able to run `examples/metadata.py` - the others try to read in data that isn't in this repo. 

![image](https://github.com/user-attachments/assets/f02fc363-1c62-4c92-b4e5-6a868089fa20)

heathhenley commented:
It has git installed already, which was nice.

I would be surprised if there were no gis (`osgeo`, `gdal`, etc?) dependencies that need to be installed via apt-get, which probably means we would want to switch to use a `Dockerfile`. You can give `Dockerfile` to `devcontainer.json` and get the benefits of caching the build, etc, but tbd if that's problem. 

The difference with `postCreateCommand` and the `Dockerfile` option is that the `Dockerfile` is run before the container is made and the workspace is mounted, so it's best for external dependencies: https://code.visualstudio.com/docs/devcontainers/create-dev-container#_dockerfile

heathhenley commented:
Successfully spun up in a "codespace" in the browser which is kind of cool

brian-r-calder commented:
Did you try using the Dockerfile.build that's in the project?  I added it for completeness, modelled on our other projects, but haven't yet attempted to spin it up in a dev container.  I kind of like the idea of having a control input that is automatic, rather than relying on a container that we have to download from elsewhere that may be subject to change.

heathhenley commented:
That makes sense, I didn't try it but I'll give it a shot and see what would need to be updated.

heathhenley commented:
I just updated this to use the existing container instead of the python one from the template library.

In that though I updated the container so that it pulls the [python 3.12 tag](https://github.com/docker-library/python/blob/9c50e071f7445d04776485e28d208d39c25fbbaa/3.12/slim-bookworm/Dockerfile) instead of installing debian 11 and then manually installing python (this seemed to be installing 3.9 by default too). It's still a debian image, but running bookworm instead of bullseye and with python 3.12 preinstalled. We could switch to use `python-3.12-slim-bullseye` if that is actually needed/preferred.

I spun it up in a "codespace" on github and it seems to work, able to import openvbi and run the metadata example.

Let me know what you think!

brian-r-calder commented:
This seems to work really well -- thanks!  The only thing that I can think of that we'd want to add is a mount into the dev container for example data, but that's either something that we'd want to install by default, or provide an example on how to do it, since it would be different for each user.  We could merge this and then add, however.

heathhenley commented:
I think that's reasonable to have in the repo IMO, I don't even think it would warrant using git lfs or anything - maybe it can be unzipped on the post create step - or just leave it with instructions to unzip if you want to use it

heathhenley commented on Dockerfile.build:
```diff
@@ -1,6 +1,7 @@
+
 
 FROM builder as final
 ENV PYTHONBUFFERED 1
```
Nice catch, just added

heathhenley commented:
One more thing - this latest version doesn't install git in the container - was not sure if it should, I've been getting a shell into the repo on my local machine to use git, because if you install it in the container you have to tell it who you are, etc to actually commit. 

If you think we should add we can install in the apt-get install at the top - I think the reason to do it would be to allow someone to completely do dev in codespaces, otherwise if they've got the dev container running they probably have git locally already configured

selimnairb commented:
> One more thing - this latest version doesn't install git in the container - was not sure if it should, I've been getting a shell into the repo on my local machine to use git, because if you install it in the container you have to tell it who you are, etc to actually commit.
> 
> If you think we should add we can install in the apt-get install at the top - I think the reason to do it would be to allow someone to completely do dev in codespaces, otherwise if they've got the dev container running they probably have git locally already configured

I would just rely on a bind mount and host Git for doing commits.
